src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -810,6 +894,162 @@ NodeListInsertCommand(List *workerNodeList)
+
+		if (!isFirstValue)
+		{
+			appendStringInfo(insertDistributedObjectsCommand, ", ");
+		}
+		isFirstValue = false;
+
+		appendStringInfo(insertDistributedObjectsCommand,
+						 "(%s, ARRAY[",
+						 quote_literal_cstr(objectType));
+
+		char *name;
+		bool firstLoop = true;
+		foreach_ptr(name, names)
+		{
+			if (!firstLoop)
+			{
+				appendStringInfo(insertDistributedObjectsCommand, ", ");
+			}
+			firstLoop = false;
+			appendStringInfoString(insertDistributedObjectsCommand, quote_literal_cstr(
+									   name));
+		}
+
+		appendStringInfo(insertDistributedObjectsCommand, "]::text[], ARRAY[");
+
+		char *arg;
+		firstLoop = true;
+		foreach_ptr(arg, args)
+		{
+			if (!firstLoop)
+			{
+				appendStringInfo(insertDistributedObjectsCommand, ", ");
+			}
+			firstLoop = false;
+			appendStringInfoString(insertDistributedObjectsCommand, quote_literal_cstr(
+									   arg));
+		}
+
+		appendStringInfo(insertDistributedObjectsCommand, "]::text[],");
+
+		if (distributionArgumentIndex == NULL)
+		{
+			appendStringInfo(insertDistributedObjectsCommand, "NULL, ");
+		}
+		else
+		{
+			appendStringInfo(insertDistributedObjectsCommand, "%d, ",
+							 *distributionArgumentIndex);
+		}
+		if (colocationId == NULL)
+		{
+			appendStringInfo(insertDistributedObjectsCommand, "NULL)");
+		}
+		else
+		{
+			appendStringInfo(insertDistributedObjectsCommand, "%d)",
+							 *colocationId);
+		}
+	}
+
+	appendStringInfo(insertDistributedObjectsCommand, ") ");
+
+	appendStringInfo(insertDistributedObjectsCommand,
+					 "SELECT citus_internal_add_object_metadata("
+					 "typetext, objnames, objargs, distargumentindex::int, colocationid::int) "
+					 "FROM distributed_object_data;");
+
+	return insertDistributedObjectsCommand->data;
+}
+
+
+/*
+ * citus_internal_add_object_metadata is an internal UDF to
+ * add a row to pg_dist_object.
+ */
+Datum
+citus_internal_add_object_metadata(PG_FUNCTION_ARGS)
+{
+	char *textType = TextDatumGetCString(PG_GETARG_DATUM(0));
+	ArrayType *nameArray = PG_GETARG_ARRAYTYPE_P(1);
+	ArrayType *argsArray = PG_GETARG_ARRAYTYPE_P(2);
+	int distributionArgumentIndexValue;
+	int colocationIdValue;
+	int *colocationId = NULL;
+	int *distributionArgumentIndex = NULL;
+
+	if (!PG_ARGISNULL(3))
+	{
+		distributionArgumentIndexValue = PG_GETARG_INT32(3);
+		distributionArgumentIndex = &distributionArgumentIndexValue;
+	}
+	if (!PG_ARGISNULL(4))
+	{
+		colocationIdValue = PG_GETARG_INT32(4);
+		colocationId = &colocationIdValue;
+	}
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* this UDF is not allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+	}
+
+	ObjectAddress objectAddress = PgGetObjectAddress(textType, nameArray, argsArray,
+													 true);
+
+	/* First, disable propagation off to not to cause infinite propagation */
+	EnableDependencyCreation = false;
+	MarkObjectDistributed(&objectAddress);
+	UpdateFunctionDistributionInfo(&objectAddress, distributionArgumentIndex,
```

Comment:
Calling `UpdateFunctionDistributionInfo` seems to have a negligible performance impact when we call this for all the objects. Can we only call this for functions?